### PR TITLE
[FLINK-21215][task] Do not overwrite the original CheckpointFailureReason in AsyncCheckpointRunnable

### PR DESCRIPTION
**Pull request by pnowojski**
Previously there were two issues:
1. Original failure reason would be hidden and replaced with CHECKPOINT_ASYNC_EXCEPTION
2. The order of cancelling AsyncCheckpointRunnable and notifing the JM was sub-optimal.
   Instead of notifing the JM first, we were first cancelling the future, which could
   create a race condition, were the secondary failure was reported first, causing
   slightly less readable logs/error messages.

## Verifying this change

Added unit test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java [hunk 1]
```diff
@@ -257,8 +257,8 @@ private void abortInternal(long cancelledId, CheckpointException exception) thro
                 Math.max(lastCancelledOrCompletedCheckpointId, cancelledId);
         numBarriersReceived = 0;
         controller.abortPendingCheckpoint(cancelledId, exception);
-        allBarriersReceivedFuture.completeExceptionally(exception);
         notifyAbort(cancelledId, exception);
+        allBarriersReceivedFuture.completeExceptionally(exception);
```
Comment (Myasuka) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java:261]: I think this line change should be in another commit compared with below changes to ensure exception is not overridden by `CHECKPOINT_ASYNC_EXCEPTION`.

Comment (Myasuka) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java:261]: The commit to avoid exception not overridden by `CHECKPOINT_ASYNC_EXCEPTION` should also be picked into release-1.11 and release-1.12.

Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java:261]: I'd think that we should backport all fixes in this commit, so I'd be okay with leaving it as is.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnable.java [hunk 2]
```diff
@@ -263,11 +264,21 @@ private void handleExecutionException(Exception e) {
                     // Otherwise this followup exception could race the original exception in
                     // failing the task.
                     try {
+                        Optional<CheckpointException> underlyingCheckpointException =
+                                ExceptionUtils.findThrowable(
+                                        checkpointException, CheckpointException.class);
+                        // If this failure is already a CheckpointException, do not overwrite the
+                        // original CheckpointFailureReason
+                        CheckpointFailureReason reportedFailureReason =
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnable.java:272]: `underlyingCheckpointException.map(...).orElse?`

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnableTest.java [hunk 3]
```diff
@@ -89,6 +78,47 @@ private void testAsyncCheckpointException(Supplier<Boolean> isTaskRunning) {
...
+                        DoneFuture.of(SnapshotResult.empty()),
+                        ExceptionallyDoneFuture.of(new CheckpointException(originalReason)),
+                        DoneFuture.of(SnapshotResult.empty())));
+
+        final TestEnvironment environment = new TestEnvironment();
+        final AsyncCheckpointRunnable runnable =
+                createAsyncRunnable(snapshotsInProgress, environment, true);
+        runnable.run();
+
+        Assert.assertSame((environment.getCause()).getCheckpointFailureReason(), originalReason);
```
Comment (AHeise) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnableTest.java:102]: nit: Remove extra braces? (also above from where you copied)

**Review by AHeise — APPROVED** [review]
LGTM % two smaller suggestions.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java (hunk 1 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java:261]: Ok, I will split those two changes.
